### ISSUE TITLE
Worker task cannot be cancelled as expected

### ISSUE THREAD

**Issue by felix-idf** [NativeScript/NativeScript#7440]
**Environment**
Provide version numbers for the following components (information can be retrieved by running `tns info` in your project folder or by inspecting the `package.json` of the project):
 - CLI: 5.4.2
 - Cross-platform modules: 5.4.2
 - Android Runtime: 5.4.0
 - iOS Runtime: -
 - Plugin(s): -

**Describe the bug**
I try to offer the opportunity to cancel a worker's task, but I am not able to terminate the worker.

Here is what I am doing:

I start the worker script this way:
```
//main-page.js
if (global["TNS_WEBPACK"]) {
    const WorkerScript = require("nativescript-worker-loader!./worker.js");
    worker = new WorkerScript();
} else {
    worker = new Worker("./worker.js");
}

worker.onmessage = function(msg) {
    if(msg.data.isWorkerLoaded) {
        console.log("worker notify state 'loaded' -> start work");

        worker.postMessage({
            action: "start"
        });
    }
};
```

This is the content of my worker script:
```
//worker.js

//...
//notify main-thread about worker being loaded
global.postMessage({isWorkerLoaded: true});

global.onmessage = function(msg) {
    if(msg.data.action === "start") {
        for(var i = 0; i < 99999999; i++) {
            console.log(i);
        }
    } else if(msg.data.action === "cancel") {
        global.close();
    }
};
```

On button tap, I want to cancel the workers task and I tried two things:
1.) calling _worker.terminate();_ from the main-thread
2.) sending a message "cancel" to the worker, and when worker receives this message, i try to call _global.close();_ from the worker

Both versions do NOT work, I still see the logs from the worker task and nothing else happens.

Btw, if I change the workers task to something 'simpler', like this
```
setInterval(function () {
        console.log("Hello");
}, 500);
```
... then worker.terminate() or global.close() works as expected.

**To Reproduce**

1.) Open the sample projected provided in the link below.
2.) Tap the button "START WORKER" to start the worker itself and the task
3.) Tap the button "CANCEL WORKER" and you see that the script is still running
4.) Open worker.js in the project, uncomment the task to the one containg "setInterval" and comment the for-loop task
5.) rebuild the app and repeat steps 1-3 to see that this way, the worker gets cancelled.

**Expected behavior**
If the worker cannot receive any messages while doing some heavy stuff, then you should be able to stop the worker calling _worker.terminate()_ from the main-thread at least.

**Sample project**
[WorkerTerminateIssue.zip](https://github.com/NativeScript/NativeScript/files/3350117/WorkerTerminateIssue.zip)

**Comment by NathanaelA** [comment]
The issue is that every JS thread is still a single threaded task; so 
```
 for(var i = 0; i < 99999999; i++) {
            console.log(i);
        }
```
This is tying up your background JS thread.    It is not allowing the message "cancel" to even be processed.  That message is queued until it can.    As for the "terminate()" -- it probably is also waiting for the JS thread to jump back into the native side so that it can see the terminate flag has been set and then close out properly....

**Comment by felix-idf** [comment]
I understand that the new message "cancel" may not be processed by the worker, if the worker is very busy.

Maybe I have not enough knowledge about workers, as I've never used them before to be honest.
But I think it should be possible for the main thread to just quit the worker thread at least, no matter what task is done on the worker thread.

**Comment by edusperoni** [comment]
According to the spec ([here](https://html.spec.whatwg.org/multipage/workers.html#dom-worker-terminate) and [here](https://html.spec.whatwg.org/multipage/workers.html#terminate-a-worker)), `worker.terminate` should stop the worker immediately. If that's not working then it's probably a bug.

>When a user agent is to terminate a worker it must run the following steps in parallel with the worker's main loop (the "run a worker" processing model defined above):
>1. Set the worker's WorkerGlobalScope object's closing flag to true.
>2. If there are any tasks queued in the WorkerGlobalScope object's event loop's task queues, discard them without processing them.
>3. **Abort the script currently running in the worker.**
>4. If the worker's WorkerGlobalScope object is actually a DedicatedWorkerGlobalScope object (i.e. the worker is a dedicated worker), then empty the port message queue of the port that the worker's implicit port is entangled with.
>
>User agents may invoke the terminate a worker algorithm when a worker stops being an active needed worker and the worker continues executing even after its closing flag was set to true.

**Comment by NathanaelA** [comment]
@edusperoni - LOL, I'm not saying you are wrong -- Thanks for looking that up.   However, there are a lot of area's that NativeScript does NOT follow the web spec, because it isn't a web app.   :grinning:       

I would love if this was fixed to actually follow the web spec; but it would not surprise me if they decided it was much safer to let the worker finish its work, because they don't know if the worker is doing anything "native" and don't want to take the risk that killing a worker doesn't cause future instability of the app.    

For example; if this was using the epson print driver; terminating the worker after it opens a connection would be a terrible thing; as then the app would no longer be ever allowed to print until it was restarted...  
(Yes I realize this is a somewhat contrived example; as I would hopefully be smart enough not to terminate the worker when trying to print; but it is used as an example of where a resource grabbed and not freed can break an app).

**Comment by edusperoni** [comment]
@NathanaelA Absolutely. Sorry if it came off that way, I actually just googled the spec to check what was the expected behavior because your answer made sense to me. :sweat_smile:

I'm aware that NativeScript doesn't follow the spec for the most part, but I can see use cases where the developer may want to forcefully terminate a task. Personally I'd just do a long running task "in parts" (using setTimeout to allow for the cancel event arrive and cleanup), but that may not always be possible.

**Comment by felix-idf** [comment]
@NathanaelA I like your print driver example and I got your point and it is right that NS does not flollow the web specs everywhere.

But I think it is not a good idea to not allow killing a worker for everyone, just because there could be somebody who kills a worker in a situation where you better should not kill the worker. 😅
It is up to the developer to decide if there is any risk when trying to abort the worker task.
